### PR TITLE
chore: helper functions for radar visualization in the radar_base library

### PR DESCRIPTION
**Pull request by kapeps**
Depends on:
- [ ] https://github.com/tidewise/perception-usv_obstacle_fusion/pull/96

# What is this PR for

This is done as the fusion library uses the same code as the radar_base orogen, so to avoid code repetition, these functions are included here in radar_base.

# How I did it

# Results, How I tested

# Checklist
- [ ] Assign yourself  in the PR
- [ ] Write unit tests (when relevant)
- [ ] Run rubocop and rake tests locally
- [ ] If this PR initialize a CMAKE package please [enable styling and linting](https://github.com/tidewise/wetpaint-buildconf/blob/master/README.md#enabling-styling-and-linting-checks-for-a-cmake-package)
- [ ] Analyse if this PR modifies the UI, if so:
    - [ ] Tested Tupan's simulation (Charts) :world_map:
    - [ ] Tested Tupan's simulation (Hud) :joystick:
    - [ ] Tested ROV's simulation :joystick:

### CODE REVIEW

File: src/Helpers.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#include "Helpers.hpp"
+
+namespace radar_base {
+    namespace helper_functions {
+
+        void iterateOverSweepsAndApplyFuntion(Radar const& radar_echo,
```
Comment (doudou) [inline src/Helpers.cpp]: Define namespace functions using the namespace explicitly (and correct spelling while you're at it, Funtion -> Function), e.g.

`helper_functions::iterateOverSweepsAndApplyFunction`

This guarantees that the declaration matches the definition. In this case, you could do

~~~
using namespace radar_base;

void helper_functions::....
~~~

File: src/Helpers.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#include "Helpers.hpp"
+
+namespace radar_base {
+    namespace helper_functions {
```
Comment (doudou) [inline src/Helpers.cpp]: Just "helpers" would be enough, and it would match the name of the file.

File: src/Helpers.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            base::Angle const& yaw_correction,
+            sweepFunction const& lambda_function);
+
+        void updateLookUpTable(int num_angles,
+            int sweep_size,
+            float beam_width,
+            int output_image_size,
+            std::unique_ptr<radar_base::EchoToImageLUT>& m_lut);
+
+        bool updateLookUpTableIfNeeded(int num_angles,
```
Comment (doudou) [inline src/Helpers.hpp]: Please document functions.

File: src/Helpers.hpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#include <vector>
+namespace radar_base {
+    namespace helpers {
+
+        /**
+         * Function to be iterated inside each sweep, its inputs are the radar echo
+         * itself, the current local echo sweep index as well as the global sweep angular
+         * position.
+         */
+        using sweepFunction = std::function<void(Radar const&, int, int)>;
```
Comment (wvmcastro) [inline src/Helpers.hpp]: Give name to parameters so it is easy to one understand what data is passed to the function. Also this define "kind of" type, shouldn't it be **PascalCase** ?
```suggestion
        using SweepFunction = std::function<void(Radar const& echo, int sweep_local_index, int sweep_global_index)>;
```

File: src/Helpers.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+}
+
+void helpers::updateLookUpTable(int num_angles,
+    int sweep_size,
+    float beam_width,
+    int output_image_size,
+    std::unique_ptr<radar_base::EchoToImageLUT>& m_lut)
+{
+    m_lut.reset(new radar_base::EchoToImageLUT(num_angles,
```
Comment (wvmcastro) [inline src/Helpers.cpp]: I don't understand why we allocate the LUT in Heap, doesn't it fit in Stack?

File: src/Helpers.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        sweep_size,
+        beam_width,
+        output_image_size));
+}
+
+bool helpers::updateLookUpTableIfNeeded(int num_angles,
```
Comment (wvmcastro) [inline src/Helpers.cpp]: I think a better way would be save the LUT parameters within the LUT itself

```cpp
class EchoToImageLUT {
public:
    int num_angles_
    int sweep_size_;
    float beam_width_;
    int window_size_;
    
    EchoToImageLUT(int num_angles,
        int sweep_size,
        float beam_width,
        int window_size)
        : num_angles_(num_angles)
        , sweep_size_(sweep_size)
        , beam_width_(beam_width)
        , window_size_(window_size)
    {
        setup(sweep_size, window_size, num_angles, beam_width);
    }
};

```
and then have a member method telling if some parameters set match this LUT parameters. 

Another thing is that `range` does not influence LUT result so why are we checking for range changes to decide weather we should reset the LUT or not?

File: src/Helpers.hpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        /**
+         * Iterates over the sweeps of a radar echo and applies a
+         * user-defined function to each sweep. This function should take the radar echo,
+         * the sweep local index and the sweep global angular position as input.
+         *
+         * @param radar_echo
+         * @param yaw_correction
+         * @param lambda_function
+         */
+        void iterateOverSweepsAndApplyFunction(Radar const& radar_echo,
```
Comment (wvmcastro) [inline src/Helpers.hpp]:
```suggestion
        /**
         * For each sweep within the echo compute its global index within the frame 
         * and provides it to a user-defined function. 
         * This function should take the radar echo,
         * the sweep local index within this echo and the sweep global index as input.
         *
         * @param radar_echo
         * @param yaw_correction
         * @param lambda_function
         */
        void computeSweepGlobalIndexAndProvidesToFunction(Radar const& radar_echo,
```
```

File: src/Helpers.hpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+         * @param radar_range
+         * @param radar_echo
+         * @param beam_width
+         * @param output_image_size
+         * @param m_lut
+         *
+         * @return true if it reset the lookup table
+         * @return false if it kept the same lookup table
+         */
+        bool updateLookUpTableIfNeeded(int num_angles,
```
Comment (wvmcastro) [inline src/Helpers.hpp]: I think that when LUT parameters correspondence is made by the LUT class, this method will become unnecessary

File: src/EchoToImageLUT.hpp [hunk 9]
```diff
@@ -13,6 +13,9 @@ namespace radar_base {
         std::vector<int> m_data_index;
 
         int m_sweep_size;
+        int m_num_angles;
+        float m_beam_width;
+        int m_window_size;
```
Comment (wvmcastro) [inline src/EchoToImageLUT.hpp:18]: I would declare all those and `m_sweep_size` as const, to make it explicit that there is no way to change a LUT during its life cycle. And maybe made them public, so there is no need to create getters in the future

File: src/Helpers.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        std::copy(radar_echo.sweep_data.begin() + current_angle * radar_echo.sweep_length,
+            radar_echo.sweep_data.begin() + (current_angle + 1) * radar_echo.sweep_length,
+            world_echoes.begin() +
+                radar_echo.sweep_length * sweep_global_angular_position);
+    }
+}
+
+void helpers::drawImageFromEchoes(std::vector<uint8_t> const& world_echoes,
+    int radar_sweep_size,
```
Comment (wvmcastro) [inline src/Helpers.cpp]: sweep size could be retrieved from lut

**Review by wvmcastro — COMMENTED** [review]
There should be a pr in `drivers/orogen/radar_base` right?

**Comment by wvmcastro** [comment]
> There should be a pr in `drivers/orogen/radar_base` right?

sorry, just found

File: src/EchoToImageLUT.hpp [hunk 11]
```diff
@@ -44,11 +47,22 @@ namespace radar_base {
...
+         * @param num_angles Amount of samples contained in a radar revolution.
+         * @param sweep_size Number of measurements inside a single sweep.
+         * @param beam_width Angle between the 3dB half power points on the main lobe.
+         * @param window_size The param will create a squared frame.
+         * @return true if the LUT configuration matches
+         * @return false if the LUT configuration doesn't match
+         */
+        bool hasMatchingConfiguration(int num_angles,
+            int sweep_size,
+            float beam_width,
```
Comment (wvmcastro) [inline src/EchoToImageLUT.hpp:61]: Is `beam_width` the same as echo `step_angle` ?

File: src/EchoToImageLUT.hpp [hunk 12]
```diff
@@ -44,15 +42,31 @@ namespace radar_base {
...
+            int window_size);
 
         void linearizeRawTable(RawTable const& table);
 
         cv::Point fetchEntry(int angle, int echo_index) const;
+
+        int const lut_sweep_size;
+        int const lut_num_angles;
+        float const lut_beam_width;
+        int const lut_window_size;
```
Comment (jhonasiv) [inline src/EchoToImageLUT.hpp]: Do they really need the lut prefix? We are already inside the LUT context

File: src/Helpers.cpp [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    if (angle < 0) {
+        angle = 2 * M_PI + angle;
+    }
+
+    int start_angle_unit = round(angle / abs(radar_echo.step_angle.getRad()));
+    int angles_in_a_frame = round(2 * M_PI / abs(radar_echo.step_angle.getRad()));
+    int signal = abs(radar_echo.step_angle.getRad()) / radar_echo.step_angle.getRad();
+    for (int current_angle = 0;
+         current_angle < static_cast<int>(radar_echo.sweep_timestamps.size());
+         current_angle++) {
```
Comment (jhonasiv) [inline src/Helpers.cpp]:
```suggestion
    for (size_t current_angle = 0;
         current_angle < radar_echo.sweep_timestamps.size();
         current_angle++) {
```

File: src/Helpers.cpp [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    std::vector<uint8_t>& world_echoes)
+{
+    double angle = (radar_echo.start_heading).getRad() + yaw_correction.getRad();
+    if (angle < 0) {
+        angle = 2 * M_PI + angle;
+    }
+
+    int start_angle_unit = round(angle / abs(radar_echo.step_angle.getRad()));
+    int angles_in_a_frame = round(2 * M_PI / abs(radar_echo.step_angle.getRad()));
+    int signal = abs(radar_echo.step_angle.getRad()) / radar_echo.step_angle.getRad();
```
Comment (jhonasiv) [inline src/Helpers.cpp]: No need to change this, but the same can be done with `copysign(1, radar_echo.step_angle.getRad())`

Comment (jhonasiv) [reply src/Helpers.cpp]: Actually, isnt the step_angle always between 0-8191? How can it be negative?

Comment (kapeps) [reply src/Helpers.cpp]: step_angle is the actual angle in radians, it can be negative.

File: src/EchoToImageLUT.hpp (hunk 12 above)
Comment (kapeps) [reply src/EchoToImageLUT.hpp]: The point of using lut_ prefix was to not have ambiguous naming with the function variables, however, as @jhonasiv pointed out, it is better to add a sufix to the variables(such as "_") instead.

File: src/Helpers.cpp (hunk 14 above)
Comment (jhonasiv) [reply src/Helpers.cpp]: Ah yes, it is a base::Angle, somehow that got past me.

File: src/EchoToImageLUT.cpp [hunk 15]
```diff
@@ -77,50 +80,57 @@ void EchoToImageLUT::linearizeRawTable(RawTable const& table)
...
+bool EchoToImageLUT::hasMatchingConfiguration(int num_angles_,
+    int sweep_size_,
+    float beam_width_,
+    int window_size_)
+{
+    if (num_angles_ == num_angles && sweep_size_ == sweep_size &&
+        beam_width_ == beam_width && window_size_ == window_size) {
+        return true;
+    }
+    return false;
```
Comment (jhonasiv) [inline src/EchoToImageLUT.cpp]:
```suggestion
   return num_angles_ == num_angles && sweep_size_ == sweep_size &&
        beam_width_ == beam_width && window_size_ == window_size;
```

File: src/EchoToImageLUT.hpp [hunk 16]
```diff
@@ -44,15 +42,31 @@ namespace radar_base {
...
+
+        /**
+         * Checks if provided configuration matches the LUT.
+         *
+         * @param num_angles Amount of samples contained in a radar revolution.
+         * @param sweep_size Number of measurements inside a single sweep.
+         * @param beam_width Angle between the 3dB half power points on the main lobe.
+         * @param window_size The param will create a squared frame.
+         * @return true if the LUT configuration matches
+         * @return false if the LUT configuration doesn't match
```
Comment (jhonasiv) [inline src/EchoToImageLUT.hpp]:
```suggestion
         * @return whether the LUT configuration matches the given parameters
```

**Review by jhonasiv — APPROVED** [review]

**Review by higor-melo — APPROVED** [review]

File: src/EchoToImageLUT.hpp [hunk 17]
```diff
@@ -44,14 +45,35 @@ namespace radar_base {
...
+            int window_size);
 
         void linearizeRawTable(RawTable const& table);
 
+        /**
+         * Draws image based on a vector representing the full rotation of the radar.
+         *
+         * @param world_echoes the vector representing a full radar rotation
+         * @param radar_sweep_size the radar sweep size
+         * @param lut the echo to image look up table
```
Comment (wvmcastro) [inline src/EchoToImageLUT.hpp]:
```suggestion
```

File: src/EchoToImageLUT.hpp [hunk 18]
```diff
@@ -44,14 +45,35 @@ namespace radar_base {
...
-            int angle,
-            int echo_index,
-            int echo,
-            bool force_write = false) const;
+        void updateImage(cv::Mat& image, int angle, int echo_index, int echo) const;
+
+        /**
+         * Checks if provided configuration matches the LUT.
+         *
+         * @param num_angles Amount of samples contained in a radar revolution.
```
Comment (wvmcastro) [inline src/EchoToImageLUT.hpp]:
```suggestion
         * @param num_angles Amount of sweeps contained in a radar revolution.
```

File: src/Radar.hpp [hunk 19]
```diff
@@ -32,6 +32,18 @@ namespace radar_base {
...
             base::Angle heading,
             uint8_t* echo_data);
+
+        /**
+         * Iterates over the sweeps of a radar echo saving them into the world echoes
+         * vector.
+         *
+         * @param radar_echo the current radar echo to be saved
+         * @param yaw_correction the yaw correction to be applied into this radar echo
+         * @param world_echoes the vector representing a full radar rotation to be updated
```
Comment (wvmcastro) [inline src/Radar.hpp]:
```suggestion
         * @param world_echoes the serialized full radar rotation to be updated
```

**Review by wvmcastro — APPROVED** [review]

**Review by jhonasiv — APPROVED** [review]

**Pull request merged**